### PR TITLE
go-fuzz-build: parallelize build

### PR DESCRIPTION
**Pull request by josselin-c**
Parallelize some build steps to improve build speed.
On my particular test on a macosx system, the build time goes from 2m05
to 1m17.

### CODE REVIEW

File: go-fuzz-build/main.go [hunk 1]
```diff
@@ -71,10 +75,25 @@ func main() {
...
+	mu := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	sem := make(chan bool, Par)
 	for _, p := range goListList(pkg, "Deps") {
-		deps[p] = goListBool(p, "Standard")
-	}
+		wg.Add(1)
+		sem <- true
+		go func(p string) {
+			res := goListBool(p, "Standard")
```
Comment (dvyukov) [inline go-fuzz-build/main.go]: If `go list` invocation is the problem, then we should extract all interesting properties and once. Currently we run `go list` 4 times per-package: to check Standart, extract Import, GoFiles CgoFiles.

File: go-fuzz-build/main.go [hunk 2]
```diff
@@ -194,9 +213,20 @@ func buildInstrumentedBinary(pkg string, deps map[string]bool, lits map[Literal]
 		// Cross-compilation is not implemented.
 		copyDir(filepath.Join(GOROOT, "pkg", runtime.GOOS+"_"+runtime.GOARCH), filepath.Join(workdir, "goroot", "pkg", runtime.GOOS+"_"+runtime.GOARCH), true, nil)
 	}
+
+	wg := sync.WaitGroup{}
```
Comment (dvyukov) [inline go-fuzz-build/main.go]: `var wg sync.WaitGroup` please as more standard.
In one codebase I counted 2598 uses of `var wg sync.WaitGroup` vs 1 for `wg := sync.WaitGroup{}`.

File: go-fuzz-build/main.go [hunk 3]
```diff
@@ -325,25 +355,40 @@ func instrumentPackages(workdir string, deps map[string]bool, lits map[Literal]s
...
-			if p1 == nil {
-				p1 = &Package{name: imp}
-				pkgs[imp] = p1
+		wg.Add(1)
+		sem <- true
+		go func(pkg string) {
+			defer wg.Done()
+			defer func() { <-sem }()
+
+			impList := goListList(pkg, "Imports")
```
Comment (dvyukov) [inline go-fuzz-build/main.go]: Here we only parallelize `go list`, right? So if we call `go list` once to extract all properties, this part will go away.

File: go-fuzz-build/main.go [hunk 4]
```diff
@@ -362,7 +407,11 @@ func instrumentPackages(workdir string, deps map[string]bool, lits map[Literal]s
 			}
 			path := filepath.Join(workdir, root, "src", p.name)
 			var files []*ast.File
-			for _, fn := range append(goListList(p.name, "GoFiles"), goListList(p.name, "CgoFiles")...) {
+			xgoFiles := make(chan []string, 2)
+			go func() { xgoFiles <- goListList(p.name, "GoFiles") }()
```
Comment (dvyukov) [inline go-fuzz-build/main.go]: This will also go away if we call `go list` once.

File: go-fuzz-build/main.go [hunk 5]
```diff
@@ -71,10 +75,25 @@ func main() {
 	}
 
 	deps := make(map[string]bool)
+	mu := sync.Mutex{}
```
Comment (dvyukov) [inline go-fuzz-build/main.go]: Similarly, `var my sync.Mutex`.

**Comment by josselin-c** [comment]
Thanks for your feedback. I'll try to find some free time to improve on this.

**Comment by AlekSi** [comment]
@josselin-c Please note that go-fuzz repository history was rewritten to reduce total repository size and download time. The easiest migration path for this PR is to `git cherry-pick` commits on top of current `master`.

**Comment by josharian** [comment]
#211 substantively changed go-fuzz-build, making it faster in the process. I can imagine that you don't want to revisit this, but it might be interesting to look at the new go-fuzz-build, and see whether your optimizations still apply.

**Comment by josharian** [comment]
I looked through the relevant commit, and found that all but one of the improvements have been incorporated into go-fuzz, by way of using go/packages. The outstanding one is parallelizing package cloning. In my experiments, that is not a significant build time factor. So I'm going to close this PR. Thanks again for taking the time to work on it.

**Pull request closed without merging**